src/Chapter3.hs

```diff
@@ -479,6 +506,34 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
+data CityBuilding = Church|Library
+newtype House = House Int
+newtype CastleName = CastleName String
+data CityType = Small | Castled String | Walled String
+data City = City
+      { citytype::CityType
+      , building:: CityBuilding
+      , houses:: [House]}
+
+
+buildCastle :: City -> String -> City
+buildCastle city name = case citytype city of
+  Small -> city {citytype=Castled name}
+  Walled _ -> city {citytype=Walled name}
+  Castled _ -> city {citytype=Castled name}
+
+buildHouse:: City -> Int -> City
+buildHouse city size = city {houses = House size :houses city}
```

Comment:
Here you pass `Int` as an argument, however, according to the task, you can only have between 1 and 4 people in the house. But here you can provide `-10` or `42`.

If you want to keep the `newtype` wrapper around `Int`, you can use an approach called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂

Reply:
Right
I did think about it, but I didn't line any of the alternatives I could think of.
House1|House2|House3|House4 isn't very elegant
smart constructor would require to put the real constructor in another module, which I understand goes in another file

Reply:
The solution with `House1 | House2 | House3 | House4` is actually pretty elegant and idiomatic here 👍🏻 
This pattern is called "making illegal states unrepresantable". We also cover it in our mini-patterns handbook:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

This particular case of data types is covered in more detail in another separate blog post if you want to learn more behind the reasoning of this approach:

* https://lexi-lambda.github.io/blog/2020/11/01/names-are-not-type-safety/

Reply:
Thank you!